src/hdmf/backends/hdf5/h5tools.py

```diff
@@ -884,10 +913,10 @@ def _filler():
         # NOTE: we can ignore options['io_settings'] for scalar data
         elif self.__is_ref(options['dtype']):
             _dtype = self.__dtypes.get(options['dtype'])
+            self.__set_written(builder)
```

Comment:
all branches of the if statement below had `builder.written = True` so I moved it to before the if statement

Reply:
The written flag should be set after the if/elif statements since that is when the state is modified. E.g., if an error occurs in ``require_dataset`` then the builder is falsely recorded as written.

Reply:
Ah, that makes sense.